lib/asciidoctor/syntax_highlighter.rb

```diff
@@ -0,0 +1,80 @@
+module Asciidoctor
+  module SyntaxHighlighter
+    PROVIDED = RUBY_ENGINE_OPAL ? {} : {
+      'coderay' => 'syntax_highlighter/coderay',
+      'highlight.js' => 'syntax_highlighter/highlightjs',
+      'highlightjs' => 'syntax_highlighter/highlightjs',
+      'html-pipeline' => 'syntax_highlighter/html_pipeline',
+      'prettify' => 'syntax_highlighter/prettify',
+      'pygments' => 'syntax_highlighter/pygments',
+    }
+
+    attr_reader :name
+
+    def initialize doc, opts = {}; end
+
+    def docinfo? location; end
+
+    def highlight?; end
+
+    def format node, source, language, opts
+      node.content
+    end
+
+    def write_stylesheet? doc; end
+
+    @lock = ::Mutex.new
```

Comment:
I think we should "hide" this implementation detail.
Opal does not provide a `Mutex` so the following exception will be thrown `Mutex: uninitialized constant Mutex`.

Reply:
Asciidoctor.js will need to provide a basic implementation of Mutex. Here's what would be required.

```
class Mutex
  def synchronize 
    yield
  end
  def owned?
    true
  end
end
```

I may also use this in the converter factory as well (and remove concurrent/hash since it's overkill)

Reply:
I have an idea to avoid the lock on read. (Not that this matters to Asciidoctor.js, which would never need the lock semantics).

Reply:
Avoided. So the idea is that only a single attempt is to made to lazy load a syntax highlighter. All other entrants defer.

Reply:
@Mogztter Are you willing to put in a stub for Mutex. I'm happy to add a fallback in core if you'd rather it be there.

Reply:
I can put a stub for Mutex in Asciidoctor.js (or Opal) but I just feel bad because the implementation will effectively do nothing. So we just adding a layer of abstraction.
But at the same time the other alternative is to add a special case in Asciidoctor core and that's not good either.

Reply:
There's nothing to feel bad about because JavaScript essentially has a global mutex. As long as the code in JavaScript doesn't yield to IO, it's synchronous.

Having a no-op implementation is just part of having a language that runs in so many different environments. In fact, the concurrent gem does exactly this for Ruby platforms that don't have threads. So providing an implementation for Mutex is an important part of Opal being a complete Ruby.